src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,120 @@
-        self.max_year = max_year or dt.MAXYEAR
-        for a in ['min_year', 'max_year']:
-            year = getattr(self, a)
-            if year < dt.MINYEAR:
-                raise InvalidArgument(u'%s out of range: %d < %d' % (
-                    a, year, dt.MINYEAR
-                ))
-            if year > dt.MAXYEAR:
-                raise InvalidArgument(u'%s out of range: %d > %d' % (
-                    a, year, dt.MAXYEAR
-                ))
-
-    def do_draw(self, data):
-        while True:
-            try:
-                result = dt.datetime(
-                    year=cu.centered_integer_range(
-                        data, self.min_year, self.max_year, 2000
-                    ),
-                    month=cu.integer_range(data, 1, 12),
-                    day=cu.integer_range(data, 1, 31),
-                    hour=cu.integer_range(data, 0, 24),
-                    minute=cu.integer_range(data, 0, 59),
-                    second=cu.integer_range(data, 0, 59),
-                    microsecond=cu.integer_range(data, 0, 999999)
-                )
-                if (
-                    not self.allow_naive or
-                    (self.timezones and cu.boolean(data))
-                ):
-                    result = cu.choice(data, self.timezones).localize(result)
-                return result
-
-            except (OverflowError, ValueError):
-                pass
-
-
-@defines_strategy
+from hypothesis._settings import note_deprecation
+
+__all__ = ['timezones', 'datetimes', 'dates', 'times']
+
+
+@st.cacheable
+@st.defines_strategy
+def timezones():
+    """Any timezone in the Olsen database, as a pytz tzinfo object.
+
+    This strategy minimises to UTC, or the smallest possible fixed
+    offset, and is designed for use with
+    :py:func:`hypothesis.strategies.datetimes`.
+
+    """
+    all_timezones = [pytz.timezone(tz) for tz in pytz.all_timezones]
+    static = [pytz.UTC] + sorted(
+        (t for t in all_timezones if isinstance(t, pytz.tzfile.StaticTzInfo)),
+        key=lambda tz: abs(tz.utcoffset(dt.datetime(2000, 1, 1)))
+    )
+    dynamic = [tz for tz in all_timezones if tz not in static]
+    return st.sampled_from(static + dynamic)
+
+
+def tz_args_strat(allow_naive, tz_list, name):
+    if tz_list is None:
+        tz_strat = timezones()
+    else:
+        tz_strat = st.sampled_from([
+            tz if isinstance(tz, dt.tzinfo) else pytz.timezone(tz)
+            for tz in tz_list
+        ])
+    if allow_naive or (allow_naive is None and tz_strat.is_empty):
+        tz_strat = st.none() | tz_strat
+    if tz_strat.is_empty:
+        raise InvalidArgument(
+            'Cannot create non-naive %s with no timezones allowed.' % name)
+    return tz_strat
+
+
+def convert_year_bound(val, default):
+    if val is None:
+        return default
+    try:
+        return default.replace(val)
+    except ValueError:
+        raise InvalidArgument('Invalid year')
```

Comment:
\*coughs smugly\* I quote from our newly established review guide: "error messages should always display the value that triggered the error"

Ahem. Sorry, I was just really pleased to have that thing merged. But it is also relevant here and I think would improve matters.